### ISSUE TITLE
Active node failover takes a lot of time due to expired tokens

### ISSUE THREAD

**Issue by oded-goldstein** [hashicorp/vault#2418]
I've been examining how to reduce the downtime caused by active node failover.
I've found that when the a new node takes the leadership it loads not only the active tokens but also the tokens that have expired.
The number of expired tokens will grow over time (according to creation and expiry of tokens) and will increase the time the active failover takes.
Is there any business use for the expired tokens or  can they be removed ?

For example a fresh cluster took around 200 ms between 
"core: stopping active operation" of the old leader to  "core/startClusterListener: serving cluster requests: cluster_listen_address" of the new leader.
An old cluster with a lot of expired tokens took up to 1:30 mintues.
In addition the same cluster took 600 ms after deleting the expired tokens from consul.
If there is no business use for the expired tokens I suggest it won't be something the active node loads.

**Comment by vishalnayak** [comment]
@oded-goldstein Tokens will have associated leases and Vault loads all the leases when a standby node takes over. When tokens expire, the leases are also destroyed. 

Which version of Vault are you using? Are you using tokens with limited use-counts by any chance? An earlier issue had caused this problem and its been fixed in 0.6.4. Please refer to this upgrade notes for more details: https://www.vaultproject.io/docs/install/upgrade-to-0.6.4.html.

**Comment by oded-goldstein** [comment]
@vishalnayak I'm using vault 0.64 ( previously used earlier versions but the load was created when working with 0.64). I'm using limited-use-count tokens to issue  leased secrets as described in the release notes.

I've tried the tidy operation but it didn't manage to resolve the issue (just reduce the failover time to around 30 secs.

**Comment by jefferai** [comment]
@oded-goldstein If you switch up to 0.6.5 (or 0.7.0 beta) it will give you a count (and a running count as it loads) of the number of leases being loaded on startup. It would be good to check that number against the number of entries returned by a list call against auth/token/accessors.

**Comment by jefferai** [comment]
Also can you go into more detail about exactly what you did? When you say "after deleting the expired tokens from consul" how did you delete these...via Vault's API (and if so, using which call) or directly in Consul somehow (and if so, which keys)? Was that run before or after the tidy operation?

**Comment by oded-goldstein** [comment]
@jefferai I will try switching to 0.6.5 but I prefer to see first if I can resolve this issue with 0.6.4 if possible.

I'll describe the procedure:
- load test an operation that creates a token using the cubbyhole wrapping (X-Vault-Wrap-TTL =30 )  as described in "The Cubbyhole Authentication Model" (https://www.hashicorp.com/blog/cubbyhole-authentication-principles/)

- after a couple of days noticed degradation in failover time. (~1:30 minutes)
- took a snapshot of consul
- removed all the expired tokens ( sys/expire/id/auth/token/)
- tested failover (~200ms)
- reloaded snapshot 
- run tidy operation
- tested failover (~ 30 secs)
So basically I've done those in isolation from each other. 

I've deleted the data directly from consul.

**Comment by jefferai** [comment]
@oded-goldstein You're still not answering the questions we need answered, but since you have direct access to Consul, rather than having you upgrade in order to give us the information we need, you can get info directly: How many keys are there, total, under `sys/expire/id`, and of those, how many are in `sys/expire/id/auth/token`?

How did you identity which tokens to remove directly from Consul? What is the failover time like if you revoke those from Vault using the API rather than deleting directly from Consul (which will leave behind associated leases and indexes)? How many were left in `sys/expire/id` and `sys/expire/id/auth/token` after doing this? How many were left in each after a tidy?

**Comment by oded-goldstein** [comment]
I now understand that the expire path is used for the expiration mechanism.
I've deleted all of them without distinction. 

Just for reference  the number after the tidy is: 47249 (for both paths)
I think my original question was just answered and I'd have to see how to limit the number of tokens being issued.

Thank you
